primitives/arithmetic/src/fixed.rs

```diff
@@ -0,0 +1,1311 @@
+	}
+
+	/// Returns zero.
+	fn zero() -> Self;
+
+	/// Checks if the number is zero.
+	fn is_zero(&self) -> bool;
+
+	/// Returns one.
+	fn one() -> Self;
+
+	/// Checks if the number is positive.
+	fn is_positive(self) -> bool;
+
+	/// Checks if the number is negative.
+	fn is_negative(self) -> bool;
+
+	/// Returns the integer part.
+	fn trunc(self) -> Self;
+
+	/// Returns the fractional part.
+	fn frac(self) -> Self;
+
+	/// Returns the smallest integer greater than or equal to a number.
+	fn ceil(self) -> Self;
+
+	/// Returns the largest integer less than or equal to a number.
+	fn floor(self) -> Self;
+
+	/// Returns the nearest integer to a number. Round half-way cases away from 0.0.
+	fn round(self) -> Self;
+}
+
+struct I129 {
+	value: u128,
+	negative: bool,
+}
+
+impl<N: FixedPointOperand> From<N> for I129 {
+	fn from(n: N) -> I129 {
+		if n < N::zero() {
+			let value: u128 = n.checked_neg()
+				.map(|n| n.unique_saturated_into())
+				.unwrap_or(N::max_value().unique_saturated_into().saturating_add(1));
+			I129 { value, negative: true }
+		} else {
+			I129 { value: n.unique_saturated_into(), negative: false }
+		}
+	}
+}
+
+/// Transforms an `I129` to `N` if it is possible.
+fn from_i129<N: FixedPointOperand>(n: I129) -> Option<N> {
+	let max_plus_one: u128 = N::max_value().unique_saturated_into().saturating_add(1);
+	if n.negative && N::min_value() < N::zero() && n.value == max_plus_one {
+		Some(N::min_value())
+	} else {
+		let unsigned_inner: N = n.value.try_into().ok()?;
+		let inner = if n.negative { unsigned_inner.checked_neg()? } else { unsigned_inner };
+		Some(inner)
+	}
+}
+
+/// Returns `R::max` or `R::min` depending on the operand signs.
+fn to_bound<N: FixedPointOperand, D: FixedPointOperand, R: Bounded>(n: N, m: D) -> R {
+	if (n < N::zero()) != (m < D::zero()) {
+		R::min_value()
+	} else {
+		R::max_value()
+	}
+}
+
+macro_rules! implement_fixed {
+	(
+		$name:ident,
+		$test_mod:ident,
+		$inner_type:ty,
+		$div:tt,
+	) => {
+		#[derive(Encode, Decode, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+		pub struct $name($inner_type);
+
+		impl From<$inner_type> for $name {
+			fn from(int: $inner_type) -> Self {
+				$name::from_integer(int)
+			}
+		}
+
+		impl<N: FixedPointOperand, D: FixedPointOperand> From<(N, D)> for $name {
+			fn from(r: (N, D)) -> Self {
+				$name::from_rational(r.0, r.1)
+			}
+		}
+
+		impl FixedPointNumber for $name {
+			type Inner = $inner_type;
+
+			const DIV: Self::Inner = $div;
+
+			fn from_inner(inner: Self::Inner) -> Self {
+				Self(inner)
+			}
+
+			fn into_inner(self) -> Self::Inner {
+				self.0
+			}
+
+			fn from_integer<N: UniqueSaturatedInto<Self::Inner>>(int: N) -> Self {
+				Self(int.unique_saturated_into().saturating_mul(Self::DIV))
+			}
+
+			fn checked_from_integer(int: Self::Inner) -> Option<Self> {
+				int.checked_mul(Self::DIV).map(Self)
+			}
+
+			fn trunc(self) -> Self {
+				Self(self.0 / Self::DIV * Self::DIV)
+			}
```

Comment:
either proof that these operations are safe or rethink.

Reply:
Added the proof.